src/parsers/simple-fasta-parser.h

```diff
@@ -104,7 +104,7 @@ void update_fasta_by_sequence(char *strptr, struct Fasta *f) {
 }
 
 struct Fasta **parse_fasta(char *fasta_str) {
-    struct Fasta **fastas = malloc(magic_number*sizeof (char*));
+    struct Fasta **fastas = malloc(magic_number*sizeof (**fastas));
```

Comment:
This looks a bit dangerous... but who knows, maybe it works.

Reply:
What I meant by this comment is that you are likely to have allocated something of the wrong size.

If you have a pointer

```c
  type *p;
```

that you want to allocated an array of, then you want some number `n` of `type`. The size of `type` is `sizeof *p`. So you would write

```c
  type *p = malloc(n * sizeof *p);
```

If you have a pointer to a pointer type

```c
  type **p;
```

and you want an array of that, then the elements in the array should have type `type *` (because that is what `p` says they have). So, you should allocate something of size `n * sizeof(type *)` or

```c
  type **p = malloc(n * sizeof *p);
```

(notice that the `malloc()` call is the same as it was for the other type--which makes sense since `type` was just any type).

If you instead write

```c
  type **p = malloc(n * sizeof **p);
```

you are allocating `n * sizeof(type)`. You could, of course, be lucky and `sizeof(type) == sizeof(type *)` (it does happen), but it would be luck and not stable to change.

If you want to allocated space for `magic_number` `struct Fasta *` pointers, you want `malloc(magic_number * sizeof *fastas)`. When you allocate `malloc(magic_number * sizeof **fastas)` you are allocating space for `magic_number` `struct Fasta` objects (which is not the same type, and not what you get if you access them through a `struct Fasta **` pointer.